examples/flutter_gallery/lib/demo/transformations/transformations_demo_gesture_transformable.dart

```diff
@@ -0,0 +1,551 @@
+  final double initialScale;
+  final double initialRotation;
+
+  @override GestureTransformableState createState() => GestureTransformableState();
+}
+
+// A single user event can only represent one of these gestures. The user can't
+// do multiple at the same time, which results in more precise transformations.
+enum _GestureType {
+  translate,
+  scale,
+  rotate,
+}
+
+// This is public only for access from a unit test.
+class GestureTransformableState extends State<GestureTransformable> with TickerProviderStateMixin {
+  Animation<Offset> _animation;
+  AnimationController _controller;
+  Animation<Matrix4> _animationReset;
+  AnimationController _controllerReset;
+  // The translation that will be applied to the scene (not viewport).
+  // A positive x offset moves the scene right, viewport left.
+  // A positive y offset moves the scene down, viewport up.
+  Offset _translateFromScene; // Point where a single translation began.
+  double _scaleStart; // Scale value at start of scaling gesture.
+  double _rotationStart = 0.0; // Rotation at start of rotation gesture.
+  Rect _boundaryRect;
+  Matrix4 _transform = Matrix4.identity();
+  double _currentRotation = 0.0;
+  _GestureType gestureType;
+
+  // The transformation matrix that gives the initial home position.
+  Matrix4 get _initialTransform {
+    Matrix4 matrix = Matrix4.identity();
+    if (widget.initialTranslation != null) {
+      matrix = matrixTranslate(matrix, widget.initialTranslation);
+    }
+    if (widget.initialScale != null) {
+      matrix = matrixScale(matrix, widget.initialScale);
+    }
+    if (widget.initialRotation != null) {
+      matrix = matrixRotate(matrix, widget.initialRotation, Offset.zero);
+    }
+    return matrix;
+  }
+
+  // Return the scene point at the given viewport point.
+  static Offset fromViewport(Offset viewportPoint, Matrix4 transform) {
+    // On viewportPoint, perform the inverse transformation of the scene to get
+    // where the point would be in the scene before the transformation.
+    final Matrix4 inverseMatrix = Matrix4.inverted(transform);
+    final Vector3 untransformed = inverseMatrix.transform3(Vector3(
+      viewportPoint.dx,
+      viewportPoint.dy,
+      0,
+    ));
+    return Offset(untransformed.x, untransformed.y);
+  }
+
+  // Get the offset of the current widget from the global screen coordinates.
+  static Offset getOffset(BuildContext context) {
+    final RenderBox renderObject = context.findRenderObject();
+    return renderObject.localToGlobal(Offset.zero);
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _boundaryRect = widget.boundaryRect ?? Offset.zero & widget.size;
+    _transform = _initialTransform;
+    _controller = AnimationController(
+      vsync: this,
+    );
+    _controllerReset = AnimationController(
+      vsync: this,
+    );
+  }
+
+  @override
+  Widget build (BuildContext context) {
+    if (widget.reset && _animationReset == null) {
```

Comment:
This looks like work that should be done at initStateTime() and in didUpdateWidget() (when the widget.reset flag changes).